install.sh

```diff
@@ -609,14 +689,87 @@ install_asset() (
   install "${archive_dir}/${binary}" "${destination}/"
 )
 
+# compare two semver strings. Returns 0 if version1 >= version2, 1 otherwise.
+# Note: pre-release (-) and metadata (+) are not supported.
+compare_semver() {
+    # remove leading 'v' if present
+    version1=$(echo "$1" | sed 's/^v//')
+    version2=$(echo "$2" | sed 's/^v//')
+
+    IFS=. read -r major1 minor1 patch1 <<EOF
+$version1
+EOF
+    IFS=. read -r major2 minor2 patch2 <<EOF
+$version2
+EOF
+
+    if [ "$major1" -gt "$major2" ]; then
+        return 0
+    elif [ "$major1" -lt "$major2" ]; then
+        return 1
+    fi
+
+    if [ "$minor1" -gt "$minor2" ]; then
+        return 0
+    elif [ "$minor1" -lt "$minor2" ]; then
+        return 1
+    fi
+
+    if [ "$patch1" -gt "$patch2" ]; then
+        return 0
+    elif [ "$patch1" -lt "$patch2" ]; then
+        return 1
+    fi
+
+    # versions are equal
+    return 0
+}
```

Comment:
(nitpick) coreutils sort and busybox sort at least support sort -V (version sort) and -c (check), so that can be used to check in a quite shorter function:
```suggestion
# compare two semver strings. Returns 0 if version1 >= version2, 1 otherwise.
# Note: pre-release (-) and metadata (+) are not supported.
compare_semver() {
    # remove leading 'v' if present
    local version1=${1#v}
    local version2=${2#v}
    
    # this checks $2 <= $1, so returns 0 if first argument <= second
    printf "%s\n" "$version2" "$version1" | sort -Vc 2>/dev/null
}
```

quick check done with (and now I noticed it, it also passes the test you added):
```
compare_semver v0.72.0 v0.72.0 || echo bad1
compare_semver v0.72.0 v0.73.0 && echo bad2
compare_semver v0.73.0 v0.72.0 || echo bad3
compare_semver 0.72.0 v0.72.0 || echo bad4
compare_semver 0.172.0 0.72.0 || echo bad5
```

Unfortunately it seems to think that 0.1-rc is higher than 0.1 which isn't semver-compliant (rc is before the release), but we probably don't really care here so I just left the comment.

(ah, this script doesn't seem to use any 'local', I don't know any shell which doesn't support the keyword but it's true that it's not posix so feel free to drop these. `${1#v}` is posix-sh compliant.)

Reply:
clever! I like this implementation better. I'll poke at it with a few more tests to make certain it works (I've been caught in situations where lexically sorting vs parse-and-sort appeared to be the same but weren't).

Reply:
so close! it seems like there is a quirk with older versions of alpine... probably a bug on their part? but I haven't dug:
```
❯ docker run --rm -it alpine:3.6 ash
   
/ # compare_semver() {
>     # remove leading 'v' if present
>     local version1=${1#v}
>     local version2=${2#v}
>     
>     # this checks $2 <= $1, so returns 0 if first argument <= second
>     printf "%s\n" "$version2" "$version1" | sort -Vc 2>/dev/null
> }

/ # compare_semver v0.72.0 v0.72.0 || echo bad1
bad1
```

But seems to work fine with later versions of alpine:
```
❯ docker run --rm -it alpine:latest ash

/ # compare_semver() {
>     # remove leading 'v' if present
>     local version1=${1#v}
>     local version2=${2#v}
>     
>     # this checks $2 <= $1, so returns 0 if first argument <= second
>     printf "%s\n" "$version2" "$version1" | sort -Vc 2>/dev/null
> }

/ # compare_semver v0.72.0 v0.72.0 || echo bad1
/ # 
```

I'll keep the original (verbose) implementation, but it isn't immediately clear why your approach doesn't work in these cases.

Reply:
alpine 3.6 sure is old! That version of busybox sort didn't have -V yet:
```
# sort -V
sort: unrecognized option: V
```
Checking other old alpines it's available since 3.10.

alpine 3.9 has been EOL since 2020-11-01 though so I'm not sure how much you need to care, but there is no problem with the pure shell version as far as I'm aware so let's keep it this way :)